packages/slate-react/src/components/string.tsx

```diff
@@ -59,21 +59,34 @@ const TextString = (props: { text: string; isTrailing?: boolean }) => {
   const { text, isTrailing = false } = props
 
   const ref = useRef<HTMLSpanElement>(null)
-  const forceUpdateCount = useRef(0)
 
-  if (ref.current && ref.current.textContent !== text) {
-    forceUpdateCount.current += 1
-  }
+  // This is the actual text rendering boundary where we interface with the DOM
+  // The text is not rendered as part of the virtual DOM, as since we handle basic character insertions natively,
+  // updating the DOM is not a one way dataflow anymore. What we need here is not reconciliation and diffing
+  // with previous version of the virtual DOM, but rather diffing with the actual DOM element, and replace the DOM <span> content
+  // exactly if and only if its current content does not match our current virtual DOM.
+  // Otherwise the DOM TextNode would always be replaced by React as the user types, which interferes with native text features,
+  // eg makes native spellcheck opt out from checking the text node.
 
-  // This component may have skipped rendering due to native operations being
-  // applied. If an undo is performed React will see the old and new shadow DOM
-  // match and not apply an update. Forces each render to actually reconcile.
-  return (
-    <span data-slate-string ref={ref} key={forceUpdateCount.current}>
-      {text}
-      {isTrailing ? '\n' : null}
-    </span>
-  )
+  // useLayoutEffect: updating our span before browser paint
+  useLayoutEffect(() => {
+    let textWithTrailing: string | null = text + (isTrailing ? '\n' : '')
+
+    // making sure we're not outputing "null" in the extreme case the text is nullish at runtime
+    if (text == null) {
+      textWithTrailing = isTrailing ? '\n' : null
+    }
```

Comment:
The comment states we want to make sure we're not outputting null, but we do have a null. Are the comment and code out of sync or am i reading it wrong?

Reply:
Sorry, I meant in the comment to avoid outputting "null" as a string, eg "null\n" as a result of `null + '\n'`.
So the intent is to set actual null in case `text == null && !isTrailing`.
Actually, we could also set empty string in this case instead of null, as it seems to two is equivalent (both will actually result in no childNodes under the span and `textContent ===  ''`).
I will clarify the comment, and simplify this also in the code.

Reply:
You kind of addressed both things i wanted to ask there, looks good!